test/concepts/test_awaitables.cpp

```diff
@@ -0,0 +1,162 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <catch2/catch.hpp>
+#include <execution.hpp>
+
+#include <coroutine.hpp>
+#include <tuple>
+#include <variant>
+
+namespace ex = std::execution;
+
+template <typename Awaiter>
+struct promise {
+  __coro::coroutine_handle<promise> get_return_object() {
+    return {__coro::coroutine_handle<promise>::from_promise(*this)};
+  }
+  __coro::suspend_always initial_suspend() noexcept { return {}; }
+  __coro::suspend_always final_suspend() noexcept { return {}; }
+  void return_void() {}
+  void unhandled_exception() {}
+
+  template <typename... T>
+  auto await_transform(T&&...) noexcept {
+    return Awaiter{};
+  }
+};
+
+struct awaiter {
+  bool await_ready() { return true; }
+  bool await_suspend(__coro::coroutine_handle<>) { return false; }
+  bool await_resume() { return false; }
+};
+
+template <typename Awaiter>
+struct awaitable_sender_1 {
+  Awaiter operator co_await();
+};
+
+struct awaitable_sender_2 {
+  using promise_type = promise<__coro::suspend_always>;
+};
+
+struct awaitable_sender_3 {
+  using promise_type = promise<awaiter>;
+};
+
+template <typename Signatures, typename Awaiter>
+void test_awaitable_sender1(Signatures&&, Awaiter&&) {
+  static_assert(ex::sender<awaitable_sender_1<Awaiter>>);
+  static_assert(ex::__awaitable<awaitable_sender_1<Awaiter>>);
+
+  awaitable_sender_1<Awaiter> s;
+  static_assert(!ex::__get_completion_signatures::__with_member_alias<awaitable_sender_1<Awaiter>>);
+  static_assert(std::is_same_v<decltype(ex::get_completion_signatures(s)), Signatures>);
+  static_assert(
+      std::is_same_v<ex::completion_signatures_of_t<awaitable_sender_1<Awaiter>>, Signatures>);
+
+  using value_types_of = ex::value_types_of_t<awaitable_sender_1<Awaiter>>;
+  using value_types = typename Signatures::template __gather_sigs<ex::set_value_t,
+      ex::__q<ex::__decayed_tuple>, ex::__q<ex::__variant>>;
+
+  static_assert(std::is_same_v<value_types_of, value_types>);
+
+  using error_types_of = ex::error_types_of_t<awaitable_sender_1<Awaiter>>;
+  using error_types = typename Signatures::template __gather_sigs<ex::set_error_t,
+      ex::__q1<ex::__id>, ex::__q<ex::__variant>>;
+
+  static_assert(std::is_same_v<error_types_of, error_types>);
+}
+
+template <typename Signatures>
+void test_awaitable_sender2(Signatures) {
+  // is_sender_v relies on get_completion_signatures and is not true
+  // even if a sender is an awaitable if it's promise type is not
+  // used to evaluate the concept awaitable
```

Comment:
This is the real problem with your patch. The `sender` concept is used without an environment to test whether a type has "sender-ness" even when it is only really a sender when used with a certain environment type.

For instance, imagine `Sndr` is the type of the expression `execution::read(execution::get_scheduler)` (which is a sender that reads a scheduler out of the environment and passes it to the receiver's `set_value). The type `Sndr` is only really a sender -- i.e., can only be connected to a receiver -- if that receiver's environment has a scheduler to give it. And yet `sender<Sndr>` would evaluate to true.

That's because the `read` sender hooks `get_completion_signatures` to return `dependent_completion_signatures<no_env>` when passed a `no_env`. That is basically like the sender responding to "are you a sender?" with "reply hazy. ask again later."

This is used to constrain sender adaptors to only accept things that look sender-ish, even though there isn't yet an environment available. Without that, we would have no way to reject obvious nonsense like `execution::then( emphatically-not-a-sender, fn )`.

With your patch, you won't be able to use just about _any_ sender adaptor with a type that is awaitable in only certain coroutine types.

The real solution would look a little like the current one with `dependent_completion_signatures`. I can think of an awaitable type implementing a dummy `operator co_await` that returns a known awaitable type that the library knows to mean "ask again later". Let's call it `dependent_awaitable`, and it can look like this:

```c++
struct dependent_awaitable {
  bool await_ready();
  dependent_awaitable await_resume();
};
```

Then the `__awaitable< Awaitable, no_env >` test will succeed (yay!) and we can have `get_completion_signatures` return `dependent_completion_signatures<no_env>` for such cases.

Make sense?

Think I'll code this up to be sure I'm not smoking dope.

Reply:
> This is the real problem with your patch. The `sender` concept is used without an environment to test whether a type has "sender-ness" even when it is only really a sender when used with a certain environment type.
> 
> For instance, imagine `Sndr` is the type of the expression `execution::read(execution::get_scheduler)` (which is a sender that reads a scheduler out of the environment and passes it to the receiver's `set_value`). The type `Sndr` is only really a sender -- i.e., can only be connected to a receiver -- if that receiver's environment has a scheduler to give it. And yet `sender<Sndr>` would evaluate to true.
> 
> That's because the `read` sender hooks `get_completion_signatures` to return `dependent_completion_signatures<no_env>` when passed a `no_env`. That is basically like the sender responding to "are you a sender?" with "reply hazy. ask again later" instead of "yes" or "no".
> 
> This is used to constrain sender adaptors to only accept things that look sender-ish, even though there isn't yet an environment available. Without that, we would have no way to reject obvious nonsense like `execution::then( emphatically-not-a-sender, fn )`.
> 
> With your patch, you won't be able to use just about _any_ sender adaptor with a type that is awaitable in only certain coroutine types.
> 
> The real solution would look a little like the current one with `dependent_completion_signatures`. I can think of an awaitable type implementing a dummy `operator co_await` that returns a known awaitable type that the library knows to mean "ask again later". Let's call it `dependent_awaitable`, and it can look like this:
> 
> ```c++
> struct dependent_awaitable {
>   bool await_ready();
>   void await_suspend(std::coroutine_handle<no_env>);
>   dependent_awaitable await_resume();
> };
> ```
> 
> Then the `__awaitable< Awaitable, no_env >` test will succeed (yay!) and we can have `get_completion_signatures` return `dependent_completion_signatures<no_env>` for such cases.
> 
> Make sense?
> 
> Think I'll code this up to be sure I'm not smoking dope.

Yes, it makes sense. That's really good, I couldn't think that far. Thanks for explaining in detail, appreciate it.